From 2056cd84dde73cc337aecef4745e4a4b83fff89b Mon Sep 17 00:00:00 2001
From: Aloka Dixit <quic_alokad@quicinc.com>
Date: Thu, 7 Dec 2023 14:54:36 -0800
Subject: [PATCH 3/7] hostapd: add driver capability

Extended feature flag NL80211_EXT_FEATURE_DEVICE_BW indicates that
the driver supports different parameters for the device bandwidth
compared to the operating bandwidth. Store the capability.
Modify nl80211_put_freq_params() include a pointer to the driver
context so that the capability can be verified when such parameters
are passed. Driver capabilities flags 
WPA_DRIVER_DEVICE_PARAMS_SUPPORTED is set to adevertise this feature 
to hostapd

Signed-off-by: Aloka Dixit <quic_alokad@quicinc.com>
Signed-off-by: Shivani Tambatkar <quic_stambatk@quicinc.com>
---
 src/drivers/driver_nl80211.c      | 20 +++++++++++---------
 src/drivers/driver_nl80211.h      |  1 +
 src/drivers/driver_nl80211_capa.c |  3 +++
 src/drivers/nl80211_copy.h        |  4 ++++
 4 files changed, 19 insertions(+), 9 deletions(-)

--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -185,7 +185,9 @@ static int nl80211_send_frame_cmd(struct
 				  size_t csa_offs_len, int link_id);
 static int wpa_driver_nl80211_probe_req_report(struct i802_bss *bss,
 					       int report);
-
+static int nl80211_put_freq_params(struct wpa_driver_nl80211_data *drv,
+				   struct nl_msg *msg,
+				   const struct hostapd_freq_params *freq);
 #define IFIDX_ANY -1
 
 static void add_ifidx(struct wpa_driver_nl80211_data *drv, int ifidx,
@@ -5022,9 +5024,9 @@ err:
 }
 #endif /* CONFIG_DRIVER_NL80211_QCA */
 
-
-static int nl80211_put_freq_params(struct nl_msg *msg,
-				   const struct hostapd_freq_params *freq)
+static int nl80211_put_freq_params(struct wpa_driver_nl80211_data *drv,
+				   struct nl_msg *msg,
+ 				   const struct hostapd_freq_params *freq)
 {
 	enum hostapd_hw_mode hw_mode;
 	int is_24ghz;
@@ -5416,7 +5418,7 @@ static int wpa_driver_nl80211_set_ap(voi
 		nla_nest_end(msg, ftm);
 	}
 
-	if (params->freq && nl80211_put_freq_params(msg, params->freq) < 0)
+	if (params->freq && nl80211_put_freq_params(drv, msg, params->freq) < 0)
 		goto fail;
 
 #ifdef CONFIG_IEEE80211AX
@@ -5598,7 +5600,7 @@ static int nl80211_set_channel(struct i8
 
 	msg = nl80211_bss_msg(bss, 0, set_chan ? NL80211_CMD_SET_CHANNEL :
 			      NL80211_CMD_SET_WIPHY);
-	if (!msg || nl80211_put_freq_params(msg, freq) < 0) {
+	if (!msg || nl80211_put_freq_params(drv, msg, freq) < 0) {
 		nlmsg_free(msg);
 		return -1;
 	}
@@ -6757,7 +6759,7 @@ retry:
 	os_memcpy(drv->ssid, params->ssid, params->ssid_len);
 	drv->ssid_len = params->ssid_len;
 
-	if (nl80211_put_freq_params(msg, &params->freq) < 0 ||
+	if (nl80211_put_freq_params(drv, msg, &params->freq) < 0 ||
 	    nl80211_put_beacon_int(msg, params->beacon_int))
 		goto fail;
 
@@ -10805,7 +10807,7 @@ static int nl80211_start_radar_detection
 	}
 
 	if (!(msg = nl80211_bss_msg(bss, 0, NL80211_CMD_RADAR_DETECT)) ||
-	    nl80211_put_freq_params(msg, freq) < 0) {
+	    nl80211_put_freq_params(drv, msg, freq) < 0) {
 		nlmsg_free(msg);
 		return -1;
 	}
@@ -11003,7 +11005,7 @@ nl80211_tdls_enable_channel_switch(void
 	if (!msg ||
 	    nla_put(msg, NL80211_ATTR_MAC, ETH_ALEN, addr) ||
 	    nla_put_u8(msg, NL80211_ATTR_OPER_CLASS, oper_class) ||
-	    (ret = nl80211_put_freq_params(msg, params))) {
+	    (ret = nl80211_put_freq_params(drv, msg, params))) {
 		nlmsg_free(msg);
 		wpa_printf(MSG_DEBUG, "nl80211: Could not build TDLS chan switch");
 		return ret;
@@ -11706,7 +11708,7 @@ static int nl80211_switch_channel(void *
 	if (!(msg = nl80211_bss_msg(bss, 0, NL80211_CMD_CHANNEL_SWITCH)) ||
 	    nla_put_u32(msg, NL80211_ATTR_CH_SWITCH_COUNT,
 			settings->cs_count) ||
-	    (ret = nl80211_put_freq_params(msg, &settings->freq_params)) ||
+	    (ret = nl80211_put_freq_params(drv, msg, &settings->freq_params)) ||
 	    (settings->block_tx &&
 	     nla_put_flag(msg, NL80211_ATTR_CH_SWITCH_BLOCK_TX)) ||
 	    (settings->punct_bitmap &&
@@ -12482,7 +12484,7 @@ static int nl80211_join_mesh(struct i802
 	wpa_printf(MSG_DEBUG, "nl80211: mesh join (ifindex=%d)", drv->ifindex);
 	msg = nl80211_drv_msg(drv, 0, NL80211_CMD_JOIN_MESH);
 	if (!msg ||
-	    nl80211_put_freq_params(msg, &params->freq) ||
+	    nl80211_put_freq_params(drv, msg, &params->freq) ||
 	    nl80211_put_basic_rates(msg, params->basic_rates) ||
 	    nl80211_put_mesh_id(msg, params->meshid, params->meshid_len) ||
 	    nl80211_put_beacon_int(msg, params->beacon_int) ||
--- a/src/drivers/driver_nl80211.h
+++ b/src/drivers/driver_nl80211.h
@@ -203,6 +203,7 @@ struct wpa_driver_nl80211_data {
 	unsigned int puncturing:1;
 	unsigned int qca_ap_allowed_freqs:1;
 	unsigned int connect_ext_vendor_cmd_avail:1;
+	unsigned int device_bw:1;
 
 	u32 ignore_next_local_disconnect;
 	u32 ignore_next_local_deauth;
--- a/src/drivers/driver_nl80211_capa.c
+++ b/src/drivers/driver_nl80211_capa.c
@@ -516,7 +516,6 @@ static void wiphy_info_tdls(struct wpa_d
 	}
 }
 
-
 static int ext_feature_isset(const u8 *ext_features, int ext_features_len,
 			     enum nl80211_ext_feature_index ftidx)
 {
@@ -724,6 +723,10 @@ static void wiphy_info_ext_feature_flags
 	if (ext_feature_isset(ext_features, len,
 			      NL80211_EXT_FEATURE_SAE_OFFLOAD_AP))
 		capa->flags2 |= WPA_DRIVER_FLAGS2_SAE_OFFLOAD_AP;
+
+	if (ext_feature_isset(ext_features, len,
+			      NL80211_EXT_FEATURE_DEVICE_BW))
+		info->drv->device_bw = 1;
 }
 
 
